From b39330a306aa02d43a24f7f875a5a358207b8302 Mon Sep 17 00:00:00 2001
From: kintalix jayanth <jayanthx.kintali@intel.com>
Date: Mon, 17 Nov 2025 18:35:04 +0530
Subject: [PATCH 2/2] partition extend device filtering to exclude vfat/fat32
 media

Combine iso9660 and vfat/fat32 detection into a unified function
isIso9660VFat(). Use blkid command to identify these filesystem types on
devices and their partitions. Exclude these devices from the list of
install targets in Calamares to prevent installation on read-only USB/media.

Signed-off-by: kintalix jayanth <jayanthx.kintali@intel.com>
---
 src/modules/partition/core/DeviceList.cpp | 24 +++++++++++++----------
 1 file changed, 14 insertions(+), 10 deletions(-)

diff --git a/src/modules/partition/core/DeviceList.cpp b/src/modules/partition/core/DeviceList.cpp
index 155d23d..ad79a39 100644
--- a/src/modules/partition/core/DeviceList.cpp
+++ b/src/modules/partition/core/DeviceList.cpp
@@ -48,34 +48,38 @@ hasRootPartition( Device* device )
  * The @p path should point to a device; blkid is used to check the FS type.
  */
 static bool
-blkIdCheckIso9660( const QString& path )
+blkIdCheckIso9660orVFat(const QString &path)
 {
     // If blkid fails, there's no output, but we don't care
     auto r = Calamares::System::runCommand( { "blkid", path }, std::chrono::seconds( 30 ) );
-    return r.getOutput().contains( "iso9660" );
+    auto output = r.getOutput();
+    return output.contains("iso9660")
+      || (( output.contains("vfat")
+      || output.contains("fat32")
+      || output.contains("fat")) && output.contains("CDROM"));
 }
 
 /// @brief Convenience to check if @p partition holds an iso9660 filesystem
 static bool
-blkIdCheckIso9660P( const Partition* partition )
+blkIdCheckIso9660orVFatP( const Partition* partition )
 {
-    return blkIdCheckIso9660( partition->partitionPath() );
+    return blkIdCheckIso9660orVFat( partition->partitionPath() );
 }
 
 /** @brief Check if the @p device is an iso9660 device
  *
- * An iso9660 device is **probably** a CD-ROM. If the device holds an
- * iso9660 FS, or any of its partitions do, then we call it an iso9660 device.
+ * An iso9660/vfat device is **probably** a CD-ROM. If the device holds an
+ * iso9660/vfat FS, or any of its partitions do, then we call it an iso9660/vfat device.
  */
 static bool
-isIso9660( const Device* device )
+isIso9660orVFat( const Device* device )
 {
     const QString path = device->deviceNode();
     if ( path.isEmpty() )
     {
         return false;
     }
-    if ( blkIdCheckIso9660( path ) )
+    if ( blkIdCheckIso9660orVFat( path ) )
     {
         return true;
     }
@@ -83,7 +87,7 @@ isIso9660( const Device* device )
     if ( device->partitionTable() && !device->partitionTable()->children().isEmpty() )
     {
         const auto& p = device->partitionTable()->children();
-        return std::any_of( p.cbegin(), p.cend(), blkIdCheckIso9660P );
+        return std::any_of( p.cbegin(), p.cend(), blkIdCheckIso9660orVFatP);
     }
     return false;
 }
@@ -180,7 +184,7 @@ getDevices( DeviceType which )
             cDebug() << Logger::SubEntry << "Removing device with root filesystem (/) on it" << it;
             it = removeInSafeMode( devices, it );
         }
-        else if ( writableOnly && isIso9660( *it ) )
+        else if ( writableOnly && isIso9660orVFat( *it ) )
         {
             cDebug() << Logger::SubEntry << "Removing device with iso9660 filesystem (probably a CD) on it" << it;
             it = removeInSafeMode( devices, it );
-- 
2.34.1

